Performance optimizations for Zinc

retronym opened the pull request.

retronym committed:
Avoid use of WrappedArray in favour of direct Array usage

retronym committed:
Enable the Scala 2.12 inliner

To be conservative, we limit the inliner to annotated methods from within the
current sources.

The motiviation is to improve the efficiency of inlinable utility
methods in ExtractAPI and Visit that will be addressed in the
subsequent commits.

retronym committed:
Avoid boxing in tuples in name hashing

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.x&zinc=pull/492/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/474/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: 1.x-pull/492/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/474/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@78f7658a67b2b3f597186207f1bdbfea67c0d313
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/492">pull/492/head       </A> | sbt/zinc@3d1dd67c97a0d2086d0c09d3d1098ceda9fd6393
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@7f8c185b4a522781d5b6dd21d3002a6bd8cfc08e
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@0d21ae6369a3e4054ed434902ea84b2520d21da9
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@9891f07bda2d7d44be141401ed29abe39d42a43a
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/492/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

retronym committed:
Hoist functions to reduce allocations in hot code path

retronym committed:
Use AnyRefMap

retronym committed:
Use iterators rather than foreach

This is not to improve performance, but rather to give cleaner profiles.

retronym committed:
Optimize annotation extraction

  - hand inline enteringPhase to avoid closure allocation
  - avoid looking up getter/setters for non fields
  - optimize for common case of no annotations

jvican commented:
Changes look good to me. Had a look at the CI, bincompat checks are failing. We should probably whitelist them.
  
```
[error]  * method hashTypeParameters(scala.collection.Seq)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.TypeParameter])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashTypeParameters")
[error]  * method hashAnnotations(scala.collection.Seq)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.Annotation])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashAnnotations")
[error]  * method hashParameters(scala.collection.Seq,xsbti.api.Type)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.TypeParameter],xsbti.api.Type)Unit instead of (scala.collection.Seq,xsbti.api.Type)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashParameters")
[error]  * method hashDefinitionsWithExtraHashes(scala.collection.Seq)Unit in class xsbt.api.HashAPI does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("xsbt.api.HashAPI.hashDefinitionsWithExtraHashes")
[error]  * method hashSeq(scala.collection.Seq,scala.Function1)Unit in class xsbt.api.HashAPI does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("xsbt.api.HashAPI.hashSeq")
[error]  * method hashValueParameters(scala.collection.Seq)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.ParameterList])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashValueParameters")
[error]  * method hashAnnotationArguments(scala.collection.Seq)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.AnnotationArgument])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashAnnotationArguments")
[error]  * method hashTypes(scala.collection.Seq,Boolean)Unit in class xsbt.api.HashAPI's type is different in current version, where it is (Array[xsbti.api.Type],Boolean)Unit instead of (scala.collection.Seq,Boolean)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.HashAPI.hashTypes")
[error]  * method visitTypeParameters(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.TypeParameter])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitTypeParameters")
[error]  * method visitDefinitions(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.Definition])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitDefinitions")
[error]  * method visitAnnotationArguments(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.AnnotationArgument])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitAnnotationArguments")
[error]  * method visitAnnotations(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.Annotation])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitAnnotations")
[error]  * method visitValueParameters(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.ParameterList])Unit instead of (scala.collection.Seq)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitValueParameters")
[error]  * method visitParameters(scala.collection.Seq,xsbti.api.Type)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.TypeParameter],xsbti.api.Type)Unit instead of (scala.collection.Seq,xsbti.api.Type)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitParameters")
[error]  * method visitTypes(scala.collection.Seq)Unit in class xsbt.api.Visit's type is different in current version, where it is (Array[xsbti.api.Type])Unit instead of (scala.collection.Seq)Unit
[error] filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitTypes")
```

retronym commented:
Yeah, I figured that wasn't really external API.

The PR is in a bit of messy state, I need to take some more time to rebase. I'd also like to try out your [benchmarking rig](https://github.com/sbt/zinc/blob/1.x/CONTRIBUTING.md#benchmarking-zinc) to see which (if any) commits have a positive impact on runtimes. I drove these changes by collecting profiles with flight recorder.

retronym committed:
Use Java collections

retronym commented:
Here's the report for `-prof gc` for the scala-library corpus.

The last column is the allocation in `B/op` attributable to Zinc itself. This PR reduces that by about a third.

```
num  desc                      gc.alloc.rate.norm   <value>-#1       
1    baseline, zinc disabled   3509985410.667  
2    baseline  zinc enabled    5148114788.000       1638129378
3    patch, zinc enabled       4612042982.667       1102057572
```

I'm thinking of adding a mode to the benchmark rig that runs the compiler up to the `ExtractAPI`  phase in `@Setup`, and loops over the `ExtractAPI` work in the benchmarked method. That will make it much easier to targetted profiling and tuning of Zinc. Of course, we must also validate the results in the broader context of the normal compilation run, as the benchmarks currently do.

retronym commented on internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala:
```diff
@@ -198,33 +202,62 @@ class ExtractAPI[GlobalType <: Global](
-      val associated =
-        List(b, b.getterIn(b.enclClass), b.setterIn(b.enclClass)).filter(_ != NoSymbol)
-      associated.flatMap(ss => mkAnnotations(in, ss.annotations)).distinct.toArray
+      if (b.hasGetter) {
```
This change needs the most careful review. We should probably build a version that has the old and new implementations in place that asserts they agree.

jvican replied:
Looks correct under the assumption that the compiler does not synthesize setters if there are no getters. We have a biggish community build in bloop (https://github.com/scalacenter/bloop/tree/master/build-integrations), I can try out the optimized Zinc version when this PR is ready and we assert that we don't break incremental compilation for existing projects. Unfortunately, these tests will need to be manual since there's no automated infrastructure for it aside from the scripted tests we have.

eed3si9n replied:
Is it possible to write out common cases as tests in https://github.com/sbt/zinc/blob/1.1.x/internal/compiler-bridge-test/src/test/scala/xsbt/ExtractAPISpecification.scala?

retronym commented on internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala:
```diff
@@ -61,17 +64,18 @@ class ExtractAPI[GlobalType <: Global](
+  private[this] val typeCache = new java.util.HashMap[(Symbol, Type), xsbti.api.Type]()
   // these caches are necessary for correctness
-  private[this] val structureCache = perRunCaches.newMap[Symbol, xsbti.api.Structure]()
+  private[this] val structureCache = new java.util.HashMap[Symbol, xsbti.api.Structure]()
```
The motivation for using Java maps was to avoid the cost of generalized equality and hashing. Usually I'd reach for an `AnyRefMap`, but AFAICT this code needs to cross compile to Scala 2.10, so it was easier to just use a Java Map.

retronym commented:
> I'm thinking of adding a mode to the benchmark rig that runs the compiler up to the `ExtractAPI`  phase in `@Setup`, and loops over the `ExtractAPI` work in the benchmarked method. 

It requires a few hacks, but seem to work: https://github.com/retronym/zinc/tree/topic/focus

WDYT?

```
[info] Benchmark                                                                       (_tempDir)  (zincEnabled)    Mode  Cnt           Score          Error   Units
[info] HotScalacMicroBenchmark.compile                                   /tmp/zinc-bench-baseline           true  sample   18        3888.819 ±      211.550   ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.00                     /tmp/zinc-bench-baseline           true  sample             3703.570                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.50                     /tmp/zinc-bench-baseline           true  sample             3839.885                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.90                     /tmp/zinc-bench-baseline           true  sample             4156.136                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.95                     /tmp/zinc-bench-baseline           true  sample             4680.843                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.99                     /tmp/zinc-bench-baseline           true  sample             4680.843                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.999                    /tmp/zinc-bench-baseline           true  sample             4680.843                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p0.9999                   /tmp/zinc-bench-baseline           true  sample             4680.843                  ms/op
[info] HotScalacMicroBenchmark.compile:compile·p1.00                     /tmp/zinc-bench-baseline           true  sample             4680.843                  ms/op
[info] HotScalacMicroBenchmark.compile:·gc.alloc.rate                    /tmp/zinc-bench-baseline           true  sample    6         281.356 ±       29.502  MB/sec
[info] HotScalacMicroBenchmark.compile:·gc.alloc.rate.norm               /tmp/zinc-bench-baseline           true  sample    6  1198942299.556 ±   819406.113    B/op
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Eden_Space           /tmp/zinc-bench-baseline           true  sample    6         281.810 ±       29.267  MB/sec
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Eden_Space.norm      /tmp/zinc-bench-baseline           true  sample    6  1200910416.889 ± 19759251.396    B/op
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Old_Gen              /tmp/zinc-bench-baseline           true  sample    6          30.626 ±        4.740  MB/sec
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Old_Gen.norm         /tmp/zinc-bench-baseline           true  sample    6   130446065.333 ±  9252509.827    B/op
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Survivor_Space       /tmp/zinc-bench-baseline           true  sample    6          18.590 ±        6.947  MB/sec
[info] HotScalacMicroBenchmark.compile:·gc.churn.PS_Survivor_Space.norm  /tmp/zinc-bench-baseline           true  sample    6    79081034.667 ± 24937547.953    B/op
[info] HotScalacMicroBenchmark.compile:·gc.count                         /tmp/zinc-bench-baseline           true  sample    6         311.000                 counts
[info] HotScalacMicroBenchmark.compile:·gc.time                          /tmp/zinc-bench-baseline           true  sample    6       10263.000                     ms
```

![image](https://user-images.githubusercontent.com/65551/36530991-467500a8-1808-11e8-92a4-3be3745d0980.png)

jvican commented on internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala:
```diff
@@ -198,33 +202,62 @@ class ExtractAPI[GlobalType <: Global](
+  // on non-fields.
+  private def annotations(in: Symbol, s: Symbol): Array[xsbti.api.Annotation] = {
+    val saved = phase
+    phase = currentRun.typerPhase
```
Is time-travelling with Scalac's API generally inefficient or impedes inlining? I see that you're inlining the body manually here.

retronym replied:
Does the on-the-fly compilation of the compiler interface enable the
optimiser? If not, hand inlining here is helpful.

eed3si9n replied:
https://github.com/sbt/zinc/blob/v1.1.1/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala#L350:

```scala
          compiler(sourceFiles, classpath, outputDirectory, "-nowarn" :: Nil)
```

It seems like we just pass in `"-nowarn"`.

jvican replied:
It would be interesting to add the optimiser flags in here and benchmark the bridge, WDYT?

eed3si9n replied:
I think it makes sense, especially if we can observe some effect. Is there a safe flag that would work for all Scala versions, or do we have to optimize the list of optimize flag per Scala versions?

retronym replied:
The compiler options and bugs are version dependent. I would be inclined to leave it off, at least by default. Exposing these options as config might be useful for experimentation.

retronym commented:
does the on-the-fly compilation of the compiler interface enable the
optimiser? If not, hand inlining here is helpful.

retronym committed:
Add whitelist entries for HashAPI changes

jvican reviewed (approved):
LGTM, this is awesome :+1:

dwijnand commented on build.sbt:
```diff
@@ -452,8 +452,29 @@ lazy val zincApiInfo = (project in internalPath / "zinc-apiinfo")
+         exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitAnnotations"),
+         exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitValueParameters"),
+         exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitParameters"),
+         exclude[IncompatibleMethTypeProblem]("xsbt.api.Visit.visitTypes"),
```
the "api" in "xsbt.api" makes me concerned: is this public API?

retronym replied:
I think the "api" here refers the the functionality of this code (visiting the API).

It would be possible to leave deprecated methods in place with the old signatures if needed.

jvican commented:
@dwijnand Is there something here you'd like to see address before we merge?

dwijnand reviewed (approved):
492 you're clear for landing

eed3si9n commented on build.sbt:
```diff
@@ -38,11 +38,11 @@ def commonSettings: Seq[Setting[_]] = Seq(
     val old = scalacOptions.value
     scalaBinaryVersion.value match {
-      case "2.12" => old
+      case "2.12" => old ++ List("-opt-inline-from:<sources>", "-opt:l:inline", "-Yopt-inline-heuristics:at-inline-annotated")
```
Due to https://github.com/sbt/zinc/issues/537, does this mean that incremental compilation of Zinc itself no longer works?
